File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/HistoricRequest.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/HistoricRequest.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Request;
import com.bloomberglp.blpapi.Datetime;

import java.util.Calendar;
import java.util.List;
import java.util.ArrayList;

public class HistoricRequest extends Request
{
    private final HistoricRequestElementStringArray _securities, _fields;
    private HistoricRequestElementDate _dtStart, _dtEnd;

    private HistoricRequestElementBool _adjustmentNormalElement, _adjustmentAbnormalElement, _adjustmentSplitElement;

    private HistoricRequestElementInt _maxDataPointElement;

    private HistoricRequestElementString _periodicityAdjustmentElement, _periodicityElement, _overrideOptionsElement, _pricingOptionElement;

    private enum HistDataPeriodicityAdjustment { actual, calendar, fiscal }
    private HistDataPeriodicityAdjustment _periodicityAdjustment = HistDataPeriodicityAdjustment.actual;

    public enum HistDataPeriodicity { daily, weekly, monthly, quarterly, semi_annually, yearly }
    private HistDataPeriodicity _periodicity = HistDataPeriodicity.daily;

    public enum PricingOption { price, yield }
    @SuppressWarnings("unused")
	private PricingOption _pricingOption = PricingOption.price;

    public enum OverrideOptions { closingPrice, averagePrice }
    @SuppressWarnings("unused")
	private OverrideOptions _overrideOptions = OverrideOptions.closingPrice;
    
    public HistoricRequest()
    {
        this._dtStart = new HistoricRequestElementDate("startDate");
        this._dtEnd = new HistoricRequestElementDate("endDate");
        this._securities = new HistoricRequestElementStringArray("securities");
        this._fields = new HistoricRequestElementStringArray("fields");
    }
    
    List<String> securities()
    {
    	return this._securities.values();
    }
    
    List<String> fields()
    {
    	return this._fields.values();
    }
    
    Datetime dtStart()
    {
    	return this._dtStart.getDate();
    }
    
    Datetime dtEnd()
    {
    	return this._dtEnd.getDate();
    }
    
    List<Datetime> getDates()
    {
    	List<Datetime> result = this.getDatesBeforeMaxPoints();
    	List<Datetime> truncatedResult = new ArrayList<Datetime>();
    	
    	if(this._maxDataPointElement == null)
    		truncatedResult = result;
    	else
    	{
    		if(result.size() > this._maxDataPointElement.getInt())
    		{
    			for(int i = result.size() - this._maxDataPointElement.getInt(); i < result.size(); i++)
    			{
    				truncatedResult.add(result.get(i));
    			}
    		}
    		else
    			truncatedResult = result;
    	}
    	
    	return truncatedResult;
    }
    
    private List<Datetime> getDatesBeforeMaxPoints()
    {
        List<Datetime> result = new ArrayList<Datetime>();
        
        Datetime dtToday = new Datetime();
        
        Datetime dtStart;
        if(this._dtStart.getDate() == null)
        {
        	dtStart = new Datetime();
        	dtStart.calendar().add(Calendar.YEAR, -1); //com.bemu.Session.SendRequest assures that dtStart is not null
        }
        else
        	dtStart = new Datetime(this._dtStart.getDate());

        
        Datetime dtEnd;
        if(this._dtEnd.getDate() == null)
        {
        	dtEnd = new Datetime();
        	dtEnd.calendar().add(Calendar.DAY_OF_MONTH, -1);
        }
        else
        	dtEnd = new Datetime(this._dtEnd.getDate());
        
        if (this._periodicityAdjustment == HistDataPeriodicityAdjustment.fiscal)
        {
        	dtStart.calendar().set(Calendar.DAY_OF_MONTH, 0);
        }
        
        Datetime dtCurrent = dtStart;
        do
        {
        	result.add(new Datetime(dtCurrent));
        	
        	switch(this._periodicity)
        	{
	        	case daily:
	        		dtCurrent.calendar().add(Calendar.DAY_OF_MONTH, 1);
	        		break;
	        	case weekly:
	        		dtCurrent.calendar().add(Calendar.DAY_OF_MONTH, 7);
	        		break;
	        	case monthly:
	        		dtCurrent.calendar().add(Calendar.MONTH, 1);
	        		break;
	        	case quarterly:
	        		dtCurrent.calendar().add(Calendar.MONTH, 3);
	        		break;
	        	case semi_annually:
	        		dtCurrent.calendar().add(Calendar.MONTH, 6);
	        		break;
	        	case yearly:
	        		dtCurrent.calendar().add(Calendar.YEAR, 1);
	        		break;
        	}

        	if(dtCurrent.calendar().get(Calendar.DAY_OF_WEEK) == Calendar.SATURDAY)
        		dtCurrent.calendar().add(Calendar.DAY_OF_MONTH, 2);
        	else if(dtCurrent.calendar().get(Calendar.DAY_OF_WEEK) == Calendar.SUNDAY)
        		dtCurrent.calendar().add(Calendar.DAY_OF_MONTH, 1);
        	
        } while (dtCurrent.calendar().getTimeInMillis() <= dtEnd.calendar().getTimeInMillis() && dtCurrent.calendar().getTimeInMillis() <= dtToday.calendar().getTimeInMillis());
        
        return result;
    }
    
    public String toString()
    {
    	StringBuilder result = new StringBuilder();
    	result.append("HistoricalDataRequest = {" + System.getProperty("line.separator"));

    	if(this._securities.numValues() > 0)
    		result.append(this._securities.prettyPrint(1));

        if (this._fields.numValues() > 0)
            result.append(this._fields.prettyPrint(1));

        Element[] elms = { this._dtStart, this._dtEnd, this._periodicityAdjustmentElement, this._periodicityElement, 
        		this._overrideOptionsElement, this._pricingOptionElement, this._maxDataPointElement, this._adjustmentNormalElement, 
        		this._adjustmentAbnormalElement, this._adjustmentSplitElement };
        
        for (int i = 0; i < elms.length; i++)
        {
            Element current = elms[i];

            if (current != null)
            {
				try { result.append(current.prettyPrint(1)); }
            	catch (Exception e) { }
            }
        }
        
        result.append("}" + System.getProperty("line.separator"));
        return result.toString();
    }

    public void append(String name, String elementValue) throws Exception
    {
    	if(name.equals("securities"))
    		this._securities.appendValue(elementValue);
    	
    	else if(name.equals("fields"))
    		this._fields.appendValue(elementValue);
    	
    	else
    		throw new Exception("not implemented. names are case-sensitive");
    }
    
    public void set(String name, String elementValue) throws Exception
    {
    	if(name.equals("startDate"))
    	{
    		Datetime dtStart = com.bloomberglp.blpapi.DisplayFormats.HistRef_TryParseInput(elementValue);
    		if(dtStart == null)
    			throw new Exception("startDate must be in the format yyyyMMdd");
    		else
    			this._dtStart = new HistoricRequestElementDate("startDate", dtStart);
    	}
    	else if(name.equals("endDate"))
    	{
    		Datetime dtEnd = com.bloomberglp.blpapi.DisplayFormats.HistRef_TryParseInput(elementValue);
    		if(dtEnd == null)
    			throw new Exception("endDate must be in the format yyyyMMdd");
    		else
    			this._dtEnd = new HistoricRequestElementDate("endDate", dtEnd);
    	}
    	else if(name.equals("periodicityAdjustment"))
    	{
    		if(elementValue.equals("CALENDAR"))
    			this._periodicityAdjustment = HistDataPeriodicityAdjustment.calendar;
    		
    		else if(elementValue.equals("FISCAL"))
    			this._periodicityAdjustment = HistDataPeriodicityAdjustment.fiscal;
    		
    		else if(elementValue.equals("ACTUAL"))
    			this._periodicityAdjustment = HistDataPeriodicityAdjustment.actual;
    		
    		else
    			throw new Exception("periodicityadjustment must be CALENDAR, FISCAL, or ACTUAL. Values are case-sensitive.");

            this._periodicityAdjustmentElement = new HistoricRequestElementString("periodicityAdjustment", elementValue);
    	}
    	else if(name.equals("periodicitySelection"))
    	{
    		if(elementValue.equals("WEEKLY"))
    			this._periodicity = HistDataPeriodicity.weekly;
    		
    		else if(elementValue.equals("MONTHLY"))
    			this._periodicity = HistDataPeriodicity.monthly;
    		
    		else if(elementValue.equals("QUARTERLY"))
    			this._periodicity = HistDataPeriodicity.quarterly;
    		
    		else if(elementValue.equals("SEMI_ANNUALLY"))
    			this._periodicity = HistDataPeriodicity.semi_annually;
    		
    		else if(elementValue.equals("YEARLY"))
    			this._periodicity = HistDataPeriodicity.yearly;
    		
    		else if(elementValue.equals("DAILY"))
    			this._periodicity = HistDataPeriodicity.daily;
    		
    		else
    			throw new Exception("periodicityselection must be DAILY, WEEKLY, MONTHLY, QUARTERLY, SEMI_ANNUALLY, or YEARLY. Values are case-sensitive.");

            this._periodicityElement = new HistoricRequestElementString("periodicitySelection", elementValue);
    	}
    	else if(name.equals("pricingOption"))
    	{
    		if(elementValue.equals("PRICING_OPTION_YIELD"))
    			this._pricingOption = PricingOption.yield;

    		else if(elementValue.equals("PRICING_OPTION_PRICE"))
    			this._pricingOption = PricingOption.price;
    		
    		else
    			throw new Exception("pricingoption must be either PRICING_OPTION_YIELD or PRICING_OPTION_PRICE. Values are case-sensitive.");

            this._pricingOptionElement = new HistoricRequestElementString("pricingOption", elementValue);
    	}
    	else if(name.equals("overrideOption"))
    	{
    		if(elementValue.equals("OVERRIDE_OPTION_GPA"))
                this._overrideOptions = OverrideOptions.averagePrice;
    		
    		else if(elementValue.equals("OVERRIDE_OPTION_CLOSE"))
                this._overrideOptions = OverrideOptions.closingPrice;
    		
    		else
    			throw new Exception("overrideoption must be either OVERRIDE_OPTION_GPA or OVERRIDE_OPTION_CLOSE. Values are case-sensitive.");

            this._overrideOptionsElement = new HistoricRequestElementString("overrideOption", elementValue);
    	}    	
    }
    
    public void set(String name, boolean elementValue) throws Exception
    {
    	if(name.equals("adjustmentNormal"))
            this._adjustmentNormalElement = new HistoricRequestElementBool("adjustmentNormal", elementValue);
    	
    	else if(name.equals("adjustmentAbnormal"))
            this._adjustmentAbnormalElement = new HistoricRequestElementBool("adjustmentAbnormal", elementValue);

    	else if(name.equals("adjustmentSplit"))
            this._adjustmentSplitElement = new HistoricRequestElementBool("adjustmentSplit", elementValue);
		
		else
			throw new Exception("not implemented. Names are case-sensitive.");    	
    }
    
    public void set(String name, int elementValue) throws Exception
    {
    	if(name.equals("maxDataPoints"))
            this._maxDataPointElement = new HistoricRequestElementInt("maxDataPoints", elementValue);

		else
			throw new Exception("not implemented. names are case-sensitive");
    }
    
    
    
}

File: BEmu/HistoricalDataRequest/HistoricRequest.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="HistoricalDataRequest/HistoricRequest.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.HistoricalDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class HistoricRequest : Request
    {
        private readonly HistoricRequestElementStringArray _securities, _fields;
        private HistoricRequestElementDate _dtStart, _dtEnd;

        private HistoricRequestElementBool _adjustmentNormalElement, _adjustmentAbnormalElement, _adjustmentSplitElement;

        private HistoricRequestElementInt _maxDataPointElement;

        private HistoricRequestElementString _periodicityAdjustmentElement, _periodicityElement, _overrideOptionsElement, _pricingOptionElement;

        private enum HistDataPeriodicityAdjustment { actual, calendar, fiscal }
        private HistDataPeriodicityAdjustment _periodicityAdjustment = HistDataPeriodicityAdjustment.actual;

        public enum HistDataPeriodicity { daily, weekly, monthly, quarterly, semi_annually, yearly }
        private HistDataPeriodicity _periodicity = HistDataPeriodicity.daily;

#pragma warning disable 0414 //disables the "is assigned but its value is never used" warning
        public enum PricingOption { price, yield }
        private PricingOption _pricingOption = PricingOption.price;

        public enum OverrideOptions { closingPrice, averagePrice }
        private OverrideOptions _overrideOptions = OverrideOptions.closingPrice;
#pragma warning restore 0414

        internal HistoricRequest()
        {
            this._dtStart = new HistoricRequestElementDate("startDate");
            this._dtEnd = new HistoricRequestElementDate("endDate");
            this._securities = new HistoricRequestElementStringArray("securities");
            this._fields = new HistoricRequestElementStringArray("fields");
        }

        internal List<string> Securities { get { return this._securities.Values; } }
        internal List<string> Fields { get { return this._fields.Values; } }
        internal DateTime? DtStart { get { return this._dtStart.GetDate; } }
        internal DateTime? DtEnd { get { return this._dtEnd.GetDate; } }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._securities;
                yield return this._fields;
                yield return this._dtStart;
                yield return this._dtEnd;

                if (this._adjustmentNormalElement != null) yield return this._adjustmentNormalElement;
                if (this._adjustmentAbnormalElement != null) yield return this._adjustmentAbnormalElement;
                if (this._adjustmentSplitElement != null) yield return this._adjustmentSplitElement;
                if (this._maxDataPointElement != null) yield return this._maxDataPointElement;
                if (this._periodicityAdjustmentElement != null) yield return this._periodicityAdjustmentElement;
                if (this._periodicityElement != null) yield return this._periodicityElement;
                if (this._overrideOptionsElement != null) yield return this._overrideOptionsElement;
                if (this._pricingOptionElement != null) yield return this._pricingOptionElement;
            }
        }

        public override Element this[string elementName]
        {
            get
            {
                if (this._securities.Name.ToString().ToLower() == elementName.ToLower())
                    return this._securities;
                else if (this._fields.Name.ToString().ToLower() == elementName.ToLower())
                    return this._fields;
                else
                    throw new NotImplementedException("BEmu.HistoricalDataRequest.RequestReference: public Element this[string elementName] not supported");
            }
        }

        #region Date Stream
        internal IEnumerable<DateTime> GetDates()
        {
            List<DateTime> result = this.GetDatesBeforeMaxPoints();

            if (this._maxDataPointElement == null)
                return result;
            else //if the results are capped, this returns the most recent dates
                return result.Skip(result.Count - this._maxDataPointElement.GetInt);
        }

        private List<DateTime> GetDatesBeforeMaxPoints()
        {
            List<DateTime> result = new List<DateTime>();

            DateTime dtStart = this._dtStart.GetDate.GetValueOrDefault(DateTime.Today.AddYears(-1)); //BEmu.Session.SendRequest assures that dtStart is not null
            DateTime dtEnd = this._dtEnd.GetDate.GetValueOrDefault(DateTime.Today.AddDays(-1));

            if (this._periodicityAdjustment == HistDataPeriodicityAdjustment.fiscal)
                dtStart.AddDays(-dtStart.Day); //I use the start of the month as the fiscal adjustment.  It's fake.

            DateTime dtCurrent = dtStart;
            do
            {
                result.Add(dtCurrent);

                switch (this._periodicity)
                {
                    case HistDataPeriodicity.daily:
                        dtCurrent = dtCurrent.AddDays(1);
                        break;
                    case HistDataPeriodicity.weekly:
                        dtCurrent = dtCurrent.AddDays(7);
                        break;
                    case HistDataPeriodicity.monthly:
                        dtCurrent = dtCurrent.AddMonths(1);
                        break;
                    case HistDataPeriodicity.quarterly:
                        dtCurrent = dtCurrent.AddMonths(3);
                        break;
                    case HistDataPeriodicity.semi_annually:
                        dtCurrent = dtCurrent.AddMonths(6);
                        break;
                    case HistDataPeriodicity.yearly:
                        dtCurrent = dtCurrent.AddYears(1);
                        break;
                    default:
                        throw new NotImplementedException(this._periodicity.ToString());
                }

                if (dtCurrent.DayOfWeek == DayOfWeek.Saturday)
                    dtCurrent = dtCurrent.AddDays(2);
                else if (dtCurrent.DayOfWeek == DayOfWeek.Sunday)
                    dtCurrent = dtCurrent.AddDays(1);

            } while (dtCurrent <= dtEnd && dtCurrent <= DateTime.Today);

            return result;
        }
        #endregion

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.AppendFormat("HistoricalDataRequest = {{{0}", Environment.NewLine);

            if (this._securities.NumValues > 0)
                result.Append(this._securities.PrettyPrint(1));

            if (this._fields.NumValues > 0)
                result.Append(this._fields.PrettyPrint(1));

            Element[] elms = { this._dtStart, this._dtEnd, this._periodicityAdjustmentElement, this._periodicityElement, this._overrideOptionsElement, 
                                 this._pricingOptionElement, this._maxDataPointElement, this._adjustmentNormalElement, this._adjustmentAbnormalElement, 
                                 this._adjustmentSplitElement };

            for (int i = 0; i < elms.Length; i++)
            {
                Element current = elms[i];

                if (current != null)
                    result.Append(current.PrettyPrint(1));
            }

            result.AppendLine("}");
            return result.ToString();
        }

        #region OVERRIDES
        public override void Append(string name, string elementValue)
        {
            switch (name)
            {
                case "securities":
                    this._securities.AppendValue(elementValue);
                    break;
                case "fields":
                    this._fields.AppendValue(elementValue);
                    break;
                default:
                    throw new ArgumentException(string.Format("BEmu.RequestHistoric.Append: Element name {0} not supported.  Case-sensitive.", name));
            }
        }

        public override void Set(string name, string elementValue)
        {
            DateTime dtTemp;
            switch (name)
            {
                case "startDate":
                    if (Types.DisplayFormats.HistoricalOrReferenceRequests.TryParseInput(elementValue, out dtTemp))
                        this._dtStart = new HistoricRequestElementDate("startDate", dtTemp);
                    else
                        throw new ArgumentException(string.Format("BEmu.Request.Set: Bad startDate format {0}, yyyyMMdd expected", elementValue));
                    break;
                case "endDate":
                    if (Types.DisplayFormats.HistoricalOrReferenceRequests.TryParseInput(elementValue, out dtTemp))
                        this._dtEnd = new HistoricRequestElementDate("endDate", dtTemp);
                    else
                        throw new ArgumentException(string.Format("BEmu.Request.Set: Bad endDate format {0}, yyyyMMdd expected", elementValue));
                    break;
                case "periodicityAdjustment":
                    switch (elementValue)
                    {
                        case "CALENDAR":
                            this._periodicityAdjustment = HistDataPeriodicityAdjustment.calendar;
                            break;
                        case "FISCAL":
                            this._periodicityAdjustment = HistDataPeriodicityAdjustment.fiscal;
                            break;
                        case "ACTUAL":
                            this._periodicityAdjustment = HistDataPeriodicityAdjustment.actual;
                            break;
                        default:
                            throw new ArgumentException(string.Format("BEmu.Request.Set: unknown value ({0}) for periodicityadjustment. These are case-sensitive.", elementValue));
                    }
                    this._periodicityAdjustmentElement = new HistoricRequestElementString("periodicityAdjustment", elementValue);
                    break;

                case "periodicitySelection":
                    switch (elementValue)
                    {
                        case "WEEKLY":
                            this._periodicity = HistDataPeriodicity.weekly;
                            break;
                        case "MONTHLY":
                            this._periodicity = HistDataPeriodicity.monthly;
                            break;
                        case "QUARTERLY":
                            this._periodicity = HistDataPeriodicity.quarterly;
                            break;
                        case "SEMI_ANNUALLY":
                            this._periodicity = HistDataPeriodicity.semi_annually;
                            break;
                        case "YEARLY":
                            this._periodicity = HistDataPeriodicity.yearly;
                            break;
                        case "DAILY":
                            this._periodicity = HistDataPeriodicity.daily;
                            break;
                        default:
                            throw new ArgumentException(string.Format("BEmu.Request.Set: unknown value ({0}) for periodicityselection. These are case-sensitive.", elementValue));
                    }
                    this._periodicityElement = new HistoricRequestElementString("periodicitySelection", elementValue);
                    break;

                case "pricingOption":
                    switch (elementValue)
                    {
                        case "PRICING_OPTION_YIELD":
                            this._pricingOption = PricingOption.yield;
                            break;
                        case "PRICING_OPTION_PRICE":
                            this._pricingOption = PricingOption.price;
                            break;
                        default:
                            throw new ArgumentException(string.Format("BEmu.Request.Set: unknown value ({0}) for pricingOption. These are case-sensitive.", elementValue));
                    }
                    this._pricingOptionElement = new HistoricRequestElementString("pricingOption", elementValue);
                    break;

                case "overrideOption":
                    switch (elementValue)
                    {
                        case "OVERRIDE_OPTION_GPA":
                            this._overrideOptions = OverrideOptions.averagePrice;
                            break;
                        case "OVERRIDE_OPTION_CLOSE":
                            this._overrideOptions = OverrideOptions.closingPrice;
                            break;
                        default:
                            throw new ArgumentException(string.Format("BEmu.Request.Set: unknown value ({0}) for overrideoption. These are case-sensitive.", elementValue));
                    }
                    this._overrideOptionsElement = new HistoricRequestElementString("overrideOption", elementValue);
                    break;

                default:
                    throw new ArgumentException(string.Format("BEmu.Request.Set: Element name {0} not supported. These are case-sensitive.", name));
            }
        }

        public override void Set(string name, bool elementValue)
        {
            switch (name)
            {
                case "adjustmentNormal":
                    this._adjustmentNormalElement = new HistoricRequestElementBool("adjustmentNormal", elementValue);
                    break;
                case "adjustmentAbnormal":
                    this._adjustmentAbnormalElement = new HistoricRequestElementBool("adjustmentAbnormal", elementValue);
                    break;
                case "adjustmentSplit":
                    this._adjustmentSplitElement = new HistoricRequestElementBool("adjustmentSplit", elementValue);
                    break;
                default:
                    throw new ArgumentException(string.Format("BEmu.Request.Set: Element name {0} not supported", name));
            }
        }

        public override void Set(string name, int elementValue)
        {
            switch (name)
            {
                case "maxDataPoints":
                    this._maxDataPointElement = new HistoricRequestElementInt("maxDataPoints", elementValue);
                    break;
                default:
                    throw new ArgumentException(string.Format("BEmu.Request.Set: Element name {0} not supported", name));
            }
        }
        #endregion
    }
}
